etcd/cmd/microshift-etcd/run.go

```diff
@@ -97,17 +112,66 @@ func (s *EtcdService) Run() error {
 	}
 	<-e.Server.ReadyNotify()
 
+	// Start up the defrag controller.
+	defragCtx, defragShutdown := context.WithCancel(context.Background())
+	go s.defragController(defragCtx, e.Server.Backend())
+
 	// Wait to be stopped.
 	sigTerm := make(chan os.Signal, 1)
 	signal.Notify(sigTerm, os.Interrupt, syscall.SIGTERM)
 	sig := <-sigTerm
 	klog.Infof("microshift-etcd received signal %v - stopping", sig)
 
+	// Shutdown the defrag controller.
+	defragShutdown()
+
 	e.Server.Stop()
 	<-e.Server.StopNotify()
 	return nil
 }
 
+func (s *EtcdService) defragController(ctx context.Context, be backend.Backend) error {
+	// If we were told to, go ahead and do a defragment now.
+	if s.doStartupDefrag {
+		if err := be.Defrag(); err != nil {
+			err = fmt.Errorf("initial defragmentation failed: %v", err)
+			klog.Error(err)
+			return err
+		}
+	}
+
+	// Stop the controller if defrags are disabled.
+	if s.defragCheckFreq == 0 {
+		klog.Warning("defragmentation has been disabled")
+		return nil
+	}
+
+	// This timer will check the fragmented conditions periodically.
+	timer := time.NewTimer(s.defragCheckFreq)
```

Comment:
Would `time.Ticker` be a better fit here? Or do you want to make sure that between `N defrag end` `N+1 defrag start` is almost exactly `defragCheckFreq` (so it takes length of defrag into an account)?

Reply:
Yeah, using time.Timer was intentional, in the (rare) cause that defrag takes a significant amount of time and/or if the user sets the time to something very low, I don't want to have the potential for multiple defrags going at the same time.
Using either is probably fine, but at least the time.Timer will save us from that edge case.